draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +397,342 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
-AKE keys (server_private_key, server_public_key) for use with the AKE, along with a Nh-byte oprf_seed. S can use
-the same pair of keys with multiple clients, and can opt to use multiple seeds (so long as they are
-kept consistent for each client). These steps can happen offline, i.e., before the registration phase.
+OPAQUE makes use of a structure `Envelope` to manage client credentials. This envelope holds information about its format and content for the client to obtain its authentication material.
 
-Once complete, the registration process proceeds as follows.
+OPAQUE allows applications to either provide custom client private and public keys for authentication or to generate them internally, making the application oblivious to the client's private key. Each public and private key value is an opaque byte string, specific to the AKE protocol in which OPAQUE is instantiated.
 
-## Credential Registration
+These two options are defined as the `Internal Key` and `External Key` modes.  See {{envelope-modes}} for their specifications.
+
+Credential material is used to pin key material to identities if desired. If no identity is given for a party, its value MUST default to its public key. The following types of credentials are hereby considered:
+
+- client_private_key: The encoded client private key for the AKE protocol.
+- client_public_key: The encoded client public key for the AKE protocol.
+- server_public_key: The encoded server public key for the AKE protocol.
+- client_identity: The client identity. This is an application-specific value, e.g., an e-mail address or normal account name. If not specified, it defaults to the client's public key.
+- server_identity: The server identity. This is typically a domain name, e.g., example.com.  If not specified, it defaults to the server's public key.
+  See {{identities}} for information about this identity.
+
+These values are used in `SecretCredentials` and `CleartextCredentials` structures as follows:
 
 ~~~
- Client (password, creds)            Server (server_private_key, server_public_key, credential_identifier, oprf_seed)
- --------------------------------------------------------------------
- (request, blind) = CreateRegistrationRequest(password)
+struct {
+ opaque client_private_key[Nsk];
+} SecretCredentials;
 
-                               request
-                      ------------------------->
+struct {
+ opaque server_public_key[Npk];
+ opaque server_identity<1..2^16-1>;
+ opaque client_identity<1..2^16-1>;
+} CleartextCredentials;
+~~~
 
-            (response, oprf_key) = CreateRegistrationResponse(request, server_public_key, credential_identifier, oprf_seed)
+The function CreateCleartextCredentials implements this logic:
 
-                               response
-                      <-------------------------
+~~~
+CreateCleartextCredentials(server_public_key, client_public_key, server_identity, client_identity)
```

Comment:
`CreateCleartextCredentials()`
The name might be misleading for the new reader.